src/utils.jl

```diff
@@ -12,3 +12,254 @@ if VERSION <= v"1.6"
+    new_ci.edges = MethodInstance[mi]
+    # XXX: setting this edge does not give us proper method invalidation, see
+    #      JuliaLang/julia#34962 which demonstrates we also need to "call" the kernel.
+    #      invoking `code_llvm` also does the necessary codegen, as does calling the
+    #      underlying C methods -- which GPUCompiler does, so everything Just Works.
+
+    # prepare the slots
+    new_ci.slotnames = Symbol[Symbol("#self#"), :ft, :tt]
+    new_ci.slotflags = UInt8[0x00 for i = 1:3]
+
+    # return the current world age (which is not technically the codegen world age,
+    # but works well enough for invalidation purposes)
+    push!(new_ci.code, ReturnNode(Base.get_world_counter()))
+    push!(new_ci.ssaflags, 0x00)   # Julia's native compilation pipeline (and its verifier) expects `ssaflags` to be the same length as `code`
+    push!(new_ci.codelocs, 1)   # see note below
+    new_ci.ssavaluetypes += 1
+
+    # NOTE: we keep the first entry of the original linetable, and use it for location info
+    #       on the call to check_cache. we can't not have a codeloc (using 0 causes
+    #       corruption of the back trace), and reusing the target function's info
+    #       has as advantage that we see the name of the kernel in the backtraces.
+
+    return new_ci
+end
+
+@eval function codegen_world_age(ft, tt)
+    $(Expr(:meta, :generated_only))
+    $(Expr(:meta,
+           :generated,
+           Expr(:new,
+                Core.GeneratedFunctionStub,
+                :codegen_world_age_generator,
+                Any[:methodinstance, :ft, :tt],
+                Any[],
+                @__LINE__,
+                QuoteNode(Symbol(@__FILE__)),
+                true)))
+end
+
+end
+
+## looking up method instances
+
+using Core.Compiler: retrieve_code_info, CodeInfo, MethodInstance, SSAValue, SlotNumber, ReturnNode
+using Base: _methods_by_ftype
+
+@inline function typed_signature(ft::Type, tt::Type)
+    u = Base.unwrap_unionall(tt)
+    return Base.rewrap_unionall(Tuple{ft, u.parameters...}, tt)
+end
+
+# create a MethodError from a function type
+# TODO: fix upstream
+function unsafe_function_from_type(ft::Type)
+    if isdefined(ft, :instance)
+        ft.instance
+    else
+        # HACK: dealing with a closure or something... let's do somthing really invalid,
+        #       which works because MethodError doesn't actually use the function
+        Ref{ft}()[]
+    end
+end
+
+"""
+    methodinstance(ft::Type, tt::Type, [world::UInt])
+
+Look up the method instance that corresponds to invoking the function with type `ft` with
+argument typed `tt`. If the `world` argument is specified, the look-up is static and will
+always return the same result. If the `world` argument is not specified, the look-up is
+dynamic and the returned method instance will automatically be invalidated when a relevant
+function is redefined.
+"""
+function methodinstance(ft::Type, tt::Type, world::Integer=tls_world_age())
```

Comment:
This looks like its still in GPUCompiler (https://github.com/JuliaGPU/GPUCompiler.jl/blob/c0c9ab7d2604c44dfb90a99332ba3eadad3679d7/src/jlgen.jl#L57) can we remove it from here?